### PR TITLE
New Rest/Service Areas Use

### PR DESCRIPTION
**Pull request by kdiluca**
# Issue

What issue is this PR targeting? If there is no issue that addresses the problem, please open a corresponding issue and link it here.

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by kdiluca** [comment]
Example 1:
`http://localhost:8002/route?json={"locations":[{"lat":40.175659937710805,"lon":-77.53925755620003},{"lat":40.166150074618265,"lon":-77.59659245610237}],"costing":"auto","directions_options":{"units":"miles","format":"osrm"}}&access_token=`
image
image

Example 2:

File: src/tyr/route_serializer_osrm.cc [hunk 1]
```diff
@@ -489,6 +489,12 @@ json::ArrayPtr intersections(const valhalla::DirectionsLeg::Maneuver& maneuver,
       if (curr_edge->use() == TripLeg::Use::TripLeg_Use_kFerryUse) {
         classes.push_back("ferry");
       }
+      if (curr_edge->use() == TripLeg::Use::TripLeg_Use_kRestAreaUse) {
+        classes.push_back("rest_area");
+      }
+      if (curr_edge->use() == TripLeg::Use::TripLeg_Use_kServiceAreaUse) {
+        classes.push_back("service_area");
```
Comment (danpaz) [inline src/tyr/route_serializer_osrm.cc:496]: Rather than add to `classes`, we should put these in a new property on the intersection object.

File: proto/options.proto [hunk 2]
```diff
@@ -137,14 +137,14 @@ message Options {
...
     locate = 2;
-    sources_to_targets = 3;
-    optimized_route = 4;
-    isochrone = 5;
-    trace_route = 6;
-    trace_attributes = 7;
-    height = 8;
-    transit_available = 9;
-    expansion = 10;
+    sources_to_targets = 6;
```
Comment (kevinkreiser) [inline proto/options.proto:140]: this seems unrelated to the pr

File: src/loki/worker.cc [hunk 3]
```diff
@@ -265,7 +264,7 @@ loki_worker_t::work(const std::list<zmq::message_t>& job,
     const auto& options = request.options();
 
     // check there is a valid action
-    if (!options.has_action() || actions.find(options.action()) == actions.cend()) {
+    if (!options.has_action()) {
```
Comment (kevinkreiser) [inline src/loki/worker.cc:267]: same as above. seems like you merged master into your branch imporoperly?

File: proto/options.proto (hunk 2 above)
Comment (kevinkreiser) [reply proto/options.proto:140]: looks like you reverted at least part of #2545

Comment (kevinkreiser) [reply proto/options.proto:140]: actualy i guess it looks like you reverted all of it :smile:

Comment (kdiluca) [reply proto/options.proto]: thanks for catching that

File: src/loki/worker.cc (hunk 3 above)
Comment (kdiluca) [reply src/loki/worker.cc]: @kevinkreiser after reverting my revert, I now fail in CI.  

> Processing triggers for dbus (1.12.2-1ubuntu1.2) ...
>  add-apt-repository ppa:valhalla-core/valhalla
>  Library, development headers and executables associated with the Valhalla open source routing project: github.com/valhalla
>  More info: https://launchpad.net/~valhalla-core/+archive/ubuntu/valhalla
> Error: retrieving gpg key timed out.

  Do you know how to fix this?

Comment (kevinkreiser) [reply src/loki/worker.cc]: yep thats emphemeral, looks like second run got it

File: src/tyr/route_serializer_osrm.cc [hunk 4]
```diff
@@ -422,12 +422,22 @@ json::ArrayPtr intersections(const valhalla::DirectionsLeg::Maneuver& maneuver,
...
         auto intersecting_edge = node->GetIntersectingEdge(m);
         bool routeable = intersecting_edge->IsTraversableOutbound(curr_edge->travel_mode());
         uint32_t bearing = static_cast<uint32_t>(intersecting_edge->begin_heading());
         edges.emplace_back(bearing, routeable, false, false);
+
+        if (curr_edge->use() == TripLeg::Use::TripLeg_Use_kRestAreaUse) {
+          intersection->emplace("rest_area", rest_area);
+          break;
+        } else if (curr_edge->use() == TripLeg::Use::TripLeg_Use_kServiceAreaUse) {
+          intersection->emplace("service_area", service_area);
```
Comment (danpaz) [inline src/tyr/route_serializer_osrm.cc:438]: These new properties should only be added to the response if requested. Can we use filter_attributes to conditionally include these?

File: test/gurka/test_rest_service_area.cc [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  odin::EnhancedTripLeg etl(*result.mutable_trip()->mutable_routes(0)->mutable_legs(0));
+  auto curr_edge = etl.GetCurrEdge(maneuver.begin_path_index());
+
+  ASSERT_TRUE(curr_edge);
+  EXPECT_EQ(curr_edge->use(), TripLeg::Use::TripLeg_Use_kRestAreaUse);  // BC
+
+  maneuver = result.directions().routes(0).legs(0).maneuver(4);
+  curr_edge = etl.GetCurrEdge(maneuver.begin_path_index());
+  EXPECT_EQ(curr_edge->use(), TripLeg::Use::TripLeg_Use_kServiceAreaUse);  // EF
```
Comment (danpaz) [inline test/gurka/test_rest_service_area.cc:41]: For this test, you can also serialize the response to JSON and assert that the `rest_area` and `service_area` properties are set.

**Comment by kdiluca** [comment]
@kevinkreiser would you mind taking a look at this PR regarding the feature flags as well as any suggestions for keeping an empty `rest_area` or `service_area` object in the response or should we return this a different way?  cc @danpaz

File: src/tyr/route_serializer_osrm.cc (hunk 4 above)
Comment (kevinkreiser) [reply src/tyr/route_serializer_osrm.cc:438]: ditto!

File: test/gurka/test_use.cc [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+  // Assert expected number of routes, legs, steps
+  ASSERT_EQ(d["routes"].Size(), 1);
+  ASSERT_EQ(d["routes"][0]["legs"].Size(), 1);
+  ASSERT_EQ(d["routes"][0]["legs"][0]["steps"].Size(), 3);
+  auto steps = d["routes"][0]["legs"][0]["steps"].GetArray();
+
+  // Expect maneuvers
+  gurka::assert::raw::expect_maneuvers(result, {DirectionsLeg_Maneuver_Type_kStart,
+                                                DirectionsLeg_Maneuver_Type_kLeft,
```
Comment (danpaz) [inline test/gurka/test_use.cc:113]: @dgearhart I think this shows a problem with guidance when the ramp has a rest area use -- we call out a Turn Left instead of an Exit Left. Can we use this test to figure out what changes we need to make in Odin?

File: src/mjolnir/graphbuilder.cc [hunk 7]
```diff
@@ -715,7 +715,8 @@ void BuildTileSet(const std::string& ways_file,
           // Set use to ramp or turn channel
           if (edge.attributes.turn_channel) {
             directededge.set_use(Use::kTurnChannel);
-          } else if (edge.attributes.link) {
+            // Do not overwrite rest area or service area use for ramps
+          } else if (edge.attributes.link && (use != Use::kServiceArea && use != Use::kRestArea)) {
```
Comment (danpaz) [inline src/mjolnir/graphbuilder.cc:719]: @gknisely I made this change to allow links to be marked with rest or service area use, previously a motorway_link was getting ramp use instead of rest/service use. Can you take a look, is this the right thing to do here?

File: test/gurka/test_use.cc (hunk 6 above)
Comment (dgearhart) [reply test/gurka/test_use.cc:113]: @danpaz yes, I already [commented](https://github.com/mapbox/valhalla-data/issues/215#issuecomment-690370405) that guidance code will need to be updated to work with the new `Use` values

File: src/mjolnir/graphbuilder.cc (hunk 7 above)
Comment (gknisely) [reply src/mjolnir/graphbuilder.cc:719]: seems correct.  @kdiluca can you run this on a small extract.

**Comment by kdiluca** [comment]
@danpaz  @gknisely   **Updated logic produces correct results.  However, 2 gurka tests are failing so I need to figure out why.**  

image
```http://localhost:8002/route?json={"locations":[{"lat":40.16670758611766,"lon":-77.59040225297213},{"lat":40.166543612622554,"lon":-77.5955306366086}],"costing":"auto","directions_options":{"units":"miles","format":"osrm"}}&access_token=
```
image

**Review by gknisely — APPROVED** [review]

File: src/tyr/route_serializer_osrm.cc [hunk 8]
```diff
@@ -418,19 +418,36 @@ json::ArrayPtr intersections(const valhalla::DirectionsLeg::Maneuver& maneuver,
...
     if (i > 0 && !arrive_maneuver) {
+      std::vector<std::string> rest_stop_types;
       for (uint32_t m = 0; m < node->intersecting_edge_size(); m++) {
         auto intersecting_edge = node->GetIntersectingEdge(m);
         bool routeable = intersecting_edge->IsTraversableOutbound(curr_edge->travel_mode());
-        uint32_t bearing = static_cast<uint32_t>(intersecting_edge->begin_heading());
-        edges.emplace_back(bearing, routeable, false, false);
+
+        if (routeable && intersecting_edge->use() == TripLeg_Use_kRestAreaUse) {
+          rest_stop_types.push_back("rest_area");
```
Comment (danpaz) [inline src/tyr/route_serializer_osrm.cc:429]: Don't merge yet -- have some updates to make to this serialization

**Review by danpaz — CHANGES_REQUESTED** [review]

File: src/tyr/route_serializer_osrm.cc (hunk 8 above)
Comment (kdiluca) [reply src/tyr/route_serializer_osrm.cc:429]: ok

Comment (gknisely) [reply src/tyr/route_serializer_osrm.cc:429]: cc @kdiluca

**Review by danpaz — APPROVED** [review]
